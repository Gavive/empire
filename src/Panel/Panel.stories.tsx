import React, { Component } from "react";

import styled, { ThemeProvider } from "styled-components";
import { withKnobs } from "@storybook/addon-knobs";
import { withThemes } from "storybook-addon-themes/react";
import { Props as PanelProps } from "./Panel.types";

import Panel from "./Panel";
import Button from "../Button/Button";

type User = {
  name: string;
  age: number;
  image: string;
  title: string
}

const ButtonContainer = styled.div`
  margin-top: 2em;
  margin-left: 2em;
`;

const Image = styled.img`
  margin-top: 1em;
`;

const PanelContentContainer = styled.div`
  display: flex;
  justify-content: center;
`;

const Decorator = (props) => {
  let theme = lightTheme;

  if (props.theme && props.theme.name === "dark") {
    theme = darkTheme;
  }

  return (
    <ThemeProvider theme={theme}>
      {props.children}
    </ThemeProvider>
  )
}

const EmmaBaasch: User = {
  name: "Emma Baasch",
  age: 23,
  title: "Designer 2",
  image: "https://i.ytimg.com/vi/_joUTZzb9gY/maxresdefault.jpg"
};

const NaomiMori: User = {
  name: "Naomi Mori",
  age: 24,
  title: "Product Manager 2",
  image: "https://i.ytimg.com/vi/3cEionTTxZs/maxresdefault.jpg"
};

const TiffanyWood: User = {
  name: "Tiffany Wood",
  age: 22,
  title: "Software Engineer",
  image: "https://i.pinimg.com/736x/56/33/57/563357f95034ac4ebda3cb9aaf5520d8.jpg"
};

const DaisyLawrence : User = {
  name: "Daisy Lawrence",
  age: 25,
  title: "Senior Software Engineer",
  image: "https://i.ytimg.com/vi/QwISTCaSLJA/maxresdefault.jpg"
};

export default {
  component: Panel ,
  title: "Panel",
  decorators: [withKnobs, withThemes],
  parameters: {
    layout: 'fullscreen',
    themes: {
      Decorator,
      list: [
        {
          name: "light",
          color: "#fff",
          default: true
        },
        {
          name: "dark",
          color: "#121212"
        }
      ]
    },
    controls: { expanded: true }
  }
};

const darkTheme = {
  name: "dark",
  color: "white"
};

const lightTheme = {
  name: "light",
  color: "black"
};

class StoryComponent extends Component<{ args: PanelProps }, { isOpen: boolean, user: User }> {
  constructor(props: { args: PanelProps }) {
    super(null);

    this.state = {
      isOpen: props.args.isOpen,
      user: EmmaBaasch
    };
  }

  render() {
    return (
      <React.Fragment>
        <ButtonContainer>
          <Button label={EmmaBaasch.name} onClick={() => this.setState({ isOpen: true, user: EmmaBaasch })}/>
        </ButtonContainer>
        <ButtonContainer>
          <Button label={NaomiMori.name} onClick={() => this.setState({ isOpen: true, user: NaomiMori })}/>
        </ButtonContainer>
        <ButtonContainer>
          <Button label={TiffanyWood.name} onClick={() => this.setState({ isOpen: true, user: TiffanyWood })}/>
        </ButtonContainer>
        <ButtonContainer>
          <Button label={DaisyLawrence.name} onClick={() => this.setState({ isOpen: true, user: DaisyLawrence })}/>
        </ButtonContainer>
        <Panel
          title={this.state.user.name}
          subTitle={this.state.user.title}
          isOpen={this.state.isOpen}
          onClose={() => this.setState({ isOpen: false })}
        >
          <PanelContentContainer>
            <Image src={this.state.user.image} />
          </PanelContentContainer>
        </Panel>
      </React.Fragment>
    );
  }
}


export const Default = (args) => <StoryComponent args={args} />;

Default.args = {
};
