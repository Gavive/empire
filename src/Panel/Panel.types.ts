export type Props = {
  /**
   * Title of the panel.
   */
  title?: string;

  /**
   * Sub-title for the panel under the title.
   */
  subTitle?: string;

  /**
   * Indicates if the panel is currently open or not.
   * @default false
   */
  isOpen?: boolean;

  /**
   * Callback for when the close button is clicked on the
   * panel.
   * @default undefined
   */
  onClose?: () => void;

  /**
   * Apart from using the close button on the panel, the panel
   * can also be closed by clicking or tapping the area outside the panel.
   * @default false
   */
  preventLightDismiss?: boolean;
};

export type State = {
};
