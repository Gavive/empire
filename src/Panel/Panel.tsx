import React, { Component, ReactNode, Fragment } from "react";
import styled from "styled-components";
import { Close } from "@styled-icons/evil/Close";
import { Props, State } from "./Panel.types";
import { motion } from "framer-motion";

const Overlay = styled.div`
  top: 0em;
  left: 0em;
  right: 0em;
  bottom: em;
  background-color: rgba(255, 255, 255, 0.4);
  height: 100vh;
  width: 100vw;
  position: fixed;
  z-index: 1;
`;

const PanelContainer = styled(motion.nav)`
  position: fixed;
  box-shadow: rgba(0, 0, 0, 0.22) 0px 25.6px 57.6px 0px,
              rgba(0, 0, 0, 0.18) 0px 4.8px 14.4px 0px;
  display: flex;
  flex-direction: column;
  overflow: hidden auto;
  background-color: white;
  bottom: 0em;
  top: 0em;
  left: auto;
  height: 100vh;
  width: 25em;
  right: 0em;
  z-index: 2;
`;

const Title = styled.span`
    font-family: "Segoe UI", "Segoe UI Web (West European)", "Segoe UI", -apple-system, BlinkMacSystemFont, Roboto, "Helvetica Neue", sans-serif;
    -webkit-font-smoothing: antialiased;
    font-size: 20px;
    font-weight: 600;
    color: rgb(50, 49, 48);
    line-height: 27px;
    overflow-wrap: break-word;
    word-break: break-word;
    hyphens: auto;
    float: left;
`;

const NavigationHeader = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin-top: 1em;
  margin-right: 1em;
  margin-left: 1em;
  padding: 0em;

  :after {
    content: none;
  }
  :before {
    content: none;
  }
`;

const NavigationControlSection = styled.div`
`;

const CloseIcon = styled(Close)`
  height: 1.5em;
  width: 1.5em;
  float: right;
  margin-right: 0;
  color: inherit;
`;

const CloseButton = styled.button`
  background: transparent;
  border: none;
  padding: none;

  &:hover {
    cursor: pointer;
    background: #eee;
  }
`;

const SubTitle = styled.span`
  font-family: "Segoe UI", "Segoe UI Web (West European)", "Segoe UI", -apple-system, BlinkMacSystemFont, Roboto, "Helvetica Neue", sans-serif;
  -webkit-font-smoothing: antialiased;
  font-size: 1em;
  font-weight: 300;
  margin-left: 1em;
  margin-right: 1em;
`;

export default class Panel extends Component<Props, State> {
  public render(): ReactNode {
    return (
        <Fragment>
          {
            this.props.isOpen &&
            <Overlay
              onClick={() => (
                this.props.onClose && !this.props.preventLightDismiss ?
                  this.props.onClose() :
                  undefined
              )}
            />
          }
          <PanelContainer
            animate={this.props.isOpen ? "open" : "closed"}
            variants={{
              open: { opacity: 1, x: 0, width: "25em" },
              closed: { opacity: 0, x: "100%", width: 0 },
            }}
          >
            <NavigationHeader>
              <Title>{this.props.title}</Title>
              <CloseButton onClick={() => this.props.onClose ? this.props.onClose() : undefined }>
                <CloseIcon />
              </CloseButton>
            </NavigationHeader>
            <SubTitle>{this.props.subTitle}</SubTitle>
            {this.props.children}
          </PanelContainer>
        </Fragment>
    );
  }
}
