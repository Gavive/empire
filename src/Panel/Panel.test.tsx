import React from "react";
import { render } from "@testing-library/react";

import Panel from "./Panel";
import { Props } from "./Panel.types";

describe("ðŸ‘½ Panel", () => {
  let props: Props;
  const panelTitle: string = "Test panel title";
  const panelSubTitle: string = "Test panel sub title"

  beforeEach(() => {
    props = {
      title: panelTitle,
      subTitle: panelSubTitle
    };
  });

  const renderComponent = () => render(<Panel {...props} />);

  it("to match snapshot", () => {
    const component = renderComponent();

    expect(component).toMatchSnapshot();
  })
});
