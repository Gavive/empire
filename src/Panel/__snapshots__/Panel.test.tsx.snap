// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ðŸ‘½ Panel to match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="sc-AxirZ iRkAyz"
        style="opacity: 0; width: 0px; transform: translateX(0%) translateZ(0);"
      >
        <div
          class="sc-AxhCb erffSu"
        >
          <span
            class="sc-AxiKw eetsyO"
          >
            Test panel title
          </span>
          <button
            class="sc-AxheI cXKxrQ"
          >
            <svg
              aria-hidden="true"
              class="StyledIconBase-ea9ulj-0 jVnGjM sc-AxgMl hkFpyu"
              fill="currentColor"
              focusable="false"
              viewBox="0 0 50 50"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M37.304 11.282l1.414 1.414-26.022 26.02-1.414-1.413z"
              />
              <path
                d="M12.696 11.282l26.022 26.02-1.414 1.415-26.022-26.02z"
              />
            </svg>
          </button>
        </div>
        <span
          class="sc-Axmtr ksBEEX"
        >
          Test panel sub title
        </span>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="sc-AxirZ iRkAyz"
      style="opacity: 0; width: 0px; transform: translateX(0%) translateZ(0);"
    >
      <div
        class="sc-AxhCb erffSu"
      >
        <span
          class="sc-AxiKw eetsyO"
        >
          Test panel title
        </span>
        <button
          class="sc-AxheI cXKxrQ"
        >
          <svg
            aria-hidden="true"
            class="StyledIconBase-ea9ulj-0 jVnGjM sc-AxgMl hkFpyu"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 50 50"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M37.304 11.282l1.414 1.414-26.022 26.02-1.414-1.413z"
            />
            <path
              d="M12.696 11.282l26.022 26.02-1.414 1.415-26.022-26.02z"
            />
          </svg>
        </button>
      </div>
      <span
        class="sc-Axmtr ksBEEX"
      >
        Test panel sub title
      </span>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
